tests/test_extract_contained_procedures.py

```diff
@@ -0,0 +1,594 @@
+        subroutine outer()
+            implicit none
+            use parkind1, only: jprb, jpim
+            real(kind=jprb) :: x(3)
+            integer(kind=jpim) :: ii(30)
+            ii = 72
+            x(1) = 42
+            call inner()
+            contains
+            subroutine inner()
+                integer :: y
+                integer :: z
+                y = 1
+                ii(4) = 2
+                z = x(1) + y
+            end subroutine inner
+        end subroutine outer
+    """
+    src = Sourcefile.from_source(fcode, frontend = frontend)
+    routines = extract_contained_procedures(src.routines[0])
+    assert len(routines) == 1
+    outer = src.routines[0]
+    inner = routines[0]
+    assert 'x(3)' in inner.arguments
+    assert 'ii(30)' in inner.arguments
+    call = FindNodes(CallStatement).visit(outer.body)[0]
+    kwargdict = dict(call.kwarguments)
+    assert kwargdict['x'] == 'x'
+    assert kwargdict['ii'] == 'ii'
+
+    imports = FindNodes(Import).visit(inner.spec)
+    modules = set()
+    symbols = set()
+    for imp in imports:
+        modules.add(imp.module)
+        for sym in imp.symbols:
+            symbols.add(sym)
+    assert "parkind1" in modules
+    assert len(modules) == 1
+    assert "jprb" in symbols
+    assert "jpim" in symbols
+    assert len(symbols) == 2
+
+@pytest.mark.parametrize('frontend', available_frontends())
+def test_extract_contained_procedures_kind_resolution(frontend):
+    """
+    Tests that an unresolved kind parameter in inner scope is resolved from import in outer scope.
+    """
+    fcode = """
+        subroutine outer()
+            implicit none
+            use parkind1, only: jpim
+            call inner()
+            contains
+            subroutine inner()
+                integer(kind = jpim) :: y
+                integer :: z
+                z = y
+            end subroutine inner
+        end subroutine outer
+    """
+    src = Sourcefile.from_source(fcode, frontend = frontend)
+    routines = extract_contained_procedures(src.routines[0])
+    inner = routines[0]
+    assert "jpim" in inner.import_map
+
+@pytest.mark.parametrize('frontend', available_frontends())
+def test_extract_contained_procedures_derived_type_resolution(frontend):
+    """
+    Tests that an unresolved derived type in inner scope is resolved from import in outer scope.
+    """
+    fcode = """
+        subroutine outer()
+            implicit none
+            use stuff, only: mytype
+            call inner()
+            contains
+            subroutine inner()
+                type(mytype) :: y
+                integer :: z
+                z = y%a
+            end subroutine inner
+        end subroutine outer
+    """
+    src = Sourcefile.from_source(fcode, frontend = frontend)
+    routines = extract_contained_procedures(src.routines[0])
+    inner = routines[0]
+    assert "mytype" in inner.import_map
+
+@pytest.mark.parametrize('frontend', available_frontends())
```

Comment:
For this test we have to mark OMNI as xfail (which is fine, it's a limitation of the frontend) because it is missing the `types` module:
```suggestion
@pytest.mark.parametrize('frontend', available_frontends(xfail=[(OMNI, 'Parser fails on missing types module')]))
```